storm-core/src/clj/backtype/storm/daemon/supervisor.clj

```diff
@@ -732,25 +854,90 @@
+          (setup-blob-permission conf storm-conf (.toString user-dir))
+          (doseq [local-rsrc localized-resources]
+            (let [rsrc-file-path (File. (.getFilePath local-rsrc))
+                  key-name (.getName rsrc-file-path)
+                  blob-symlink-target-name (.getName (File. (.getCurrentSymlinkPath local-rsrc)))
+                  symlink-name (get-blob-localname (get blobstore-map key-name) key-name)]
+              (create-symlink! tmproot (.getParent rsrc-file-path) symlink-name
+                blob-symlink-target-name))))
+        (catch AuthorizationException authExp
+          (log-error authExp))
+        (catch KeyNotFoundException knf
+          (log-error knf))))))
+
+(defn get-blob-file-names
+  [blobstore-map]
+  (if blobstore-map
+    (for [[k, data] blobstore-map]
+      (get-blob-localname data k))))
+
+(defn download-blobs-for-topology-succeed?
+  "Assert if all blobs are downloaded for the given topology"
+  [stormconf-path target-dir]
+  (let [storm-conf (clojurify-structure (Utils/fromCompressedJsonConf (FileUtils/readFileToByteArray (File. stormconf-path))))
+        blobstore-map (storm-conf TOPOLOGY-BLOBSTORE-MAP)
+        file-names (get-blob-file-names blobstore-map)]
+    (if (and file-names (> (count file-names) 0))
+      (every? #(Utils/checkFileExists target-dir %) file-names)
+      true)))
+
 ;; distributed implementation
 (defmethod download-storm-code
-    :distributed [conf storm-id master-code-dir supervisor download-lock]
-    ;; Downloading to permanent location is atomic
-    (let [tmproot (str (supervisor-tmp-dir conf) file-path-separator (uuid))
-          stormroot (supervisor-stormdist-root conf storm-id)
-          master-meta-file-path (master-storm-metafile-path master-code-dir)
-          supervisor-meta-file-path (supervisor-storm-metafile-path tmproot)]
-      (locking download-lock
-        (log-message "Downloading code for storm id " storm-id " from " master-code-dir)
-        (FileUtils/forceMkdir (File. tmproot))
-        (Utils/downloadFromMaster conf master-meta-file-path supervisor-meta-file-path)
-        (if (:code-distributor supervisor)
-          (.download (:code-distributor supervisor) storm-id (File. supervisor-meta-file-path)))
-        (extract-dir-from-jar (supervisor-stormjar-path tmproot) RESOURCES-SUBDIR tmproot)
-        (if (.exists (File. stormroot)) (FileUtils/forceDelete (File. stormroot)))
-        (FileUtils/moveDirectory (File. tmproot) (File. stormroot))
-        (setup-storm-code-dir conf (read-supervisor-storm-conf conf storm-id) stormroot)
-        (log-message "Finished downloading code for storm id " storm-id " from " master-code-dir))))
+  :distributed [conf storm-id master-code-dir localizer]
+  ;; Downloading to permanent location is atomic
+  (let [tmproot (str (supervisor-tmp-dir conf) file-path-separator (uuid))
+        stormroot (supervisor-stormdist-root conf storm-id)
+        blobstore (Utils/getSupervisorBlobStore conf)]
```

Comment:
I hava a small suggestions.  for example, this blobstore which is supervisor's client about blobstore maybe make me confused with real "Blobstore". can you change other name？